### ISSUE TITLE
[question] How to use older MSVC toolset

### ISSUE THREAD

**Issue by Nekto89** [conan-io/conan#8735]
Hi,

I'm sorry if this has been already answered but I couldn't find it.

Microsoft releases new minor versions of compiler every several months. How to deal with this in Conan? For CMake I'm using -Tversion=14.26 (or -Tv142,version=14.26,host=x64) to be able to use 14.26 even after updating my VS2019.
https://cmake.org/cmake/help/latest/variable/CMAKE_GENERATOR_TOOLSET.html#visual-studio-toolset-selection

I found "tools.vcvars_command" that I could probably use in my recipes for forcing specific version. Is there a way to specify minor version through settings or do I need to add this as option to every single recipe and process it accordingly?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @Nekto89 

It depends on the binary compatibility level that you want to define and maintain. For most cases, and with forward compatibility of compilers versions being ok, just upgrading the tools is enough, no need to do changes in Conan model. Upgrading VS is enough, the new compiler version will be automatically used when the build helpers call vcvars internally or by passing ``-G "VIsual Studio 16 2019"`` to cmake commands.

If you actually want to build and maintain different binaries for different minor versions of the compiler, then the first step is to start using them. In ``compiler="Visual Studio"`` you have the ``toolset`` subsetting that you could be using, and customizing more if necessary (check ``conan config install`` to maintain custom settings and sync with team and CI machines). Using the new **experimental** ``msvc`` compiler, you already have the minor versions detailed as ``"19.1", "19.10", "19.11", "19.12", "19.13", "19.14", "19.15", "19.16",``. Conan by default uses the ``19.1``, but you are free to use the detailed one that you want.

Have a look at https://docs.conan.io/en/latest/reference/config_files/settings.yml.html

The idea is that the existing build helpers, and specifically the new experimental ``conan.tools.xxxx`` (https://docs.conan.io/en/latest/reference/conanfile/tools.html) helpers will handle them accordingly to define the environment as needed.

Please let me know if this helps, thanks!

**Comment by Nekto89** [comment]
Hi @memsharded

Do I understand correctly that current recipes in conan-center won't work with experimental "msvc" compiler and need to be rewritten with new experimental helpers?

For example, 
`conan.exe install . --remote conan-center --build missing --settings arch=x86_64 --settings build_type=RelWithDebInfo --settings compiler=msvc --settings compiler.version=19.26 --settings compiler.runtime=dynamic --settings compiler.runtime_type=Release --settings compiler.cppstd=17`

```
[requires]
sqlite3/3.33.0
[generators]
cmake_find_package_multi
[options]
sqlite3:shared=False
sqlite3:build_executable=False
```

leads to cmake -G "MinGW Makefiles"

**Comment by memsharded** [comment]
Hi @Nekto89 

> Do I understand correctly that current recipes in conan-center won't work with experimental "msvc" compiler and need to be rewritten with new experimental helpers?

It depends, for consuming existing binaries, it might be good enough, because there is a compatibility fallback mode, that when you request ``msvc`` binaries, it will fallback to ``Visual Studio`` built ones, which are the ones built from conan-center-index.

If your case, is as, it seems, to build them from sources, it is very likely that it will fail. The "old" build helpers are not prepared to handle the ``msvc`` compiler, and it is unlikely that this will be implemented because of the risk of breaking, and in this case the stability guarantee is more important. So we are providing ``msvc`` support in the new ``conan.tools.xxxx`` helpers only, which are experimental and under heavy development. The ConanCenter recipes will need to transition to these new tools to prepare for Conan 2.0, but this will probably not start immediately, but wait a couple of releases more until the ``conan.tools.xxxx`` mature a bit more. If you want to start using it right now, you might need to fork and adapt the recipes from ConanCenter, and expect a bit of a bumpy ride in the next months as we stabilize all the toolchains. I know that this is a bit inconvenient and we are all looking forward to being able to use the better ``msvc`` definition, but it take a while to introduce these things if not disrupting existing users is a priority.

**Comment by Nekto89** [comment]
Where can I find example with some real library? I've tried converting one of recipes but it still doesn't want to use proper toolset. I've looked at https://github.com/conan-io/conan/blob/23ce09ad30e18872fb4caa39fd011cc1d5437cd4/conan/tools/cmake/toolchain.py#L15 and I don't see where it should convert 19.26 to CMAKE_GENERATOR_TOOLSET.

```
tc = conan.tools.cmake.CMakeToolchain(self)
print(tc.generator)
print(tc.generator_platform)
print(tc.toolset)
```
Visual Studio 16 2019
x64
None

**Comment by memsharded** [comment]
Hi @Nekto89 

If you mean to use an older compiler version from a modern VS version, then probably the changes introduced in Conan 1.35 (https://blog.conan.io/2021/04/09/New-conan-release-1-35.html) could help:

- Each ``msvc`` compiler version is by default using their canonical "VS" version (the one they were released first with).
- It is now possible to configure with:
```
[conf]
tools.microsoft.msbuild:vs_version=16
```
The VS version that you want to use. The ``CMakeToolchain`` will manage to define the necessary toolset automatically.
Is this the feature you need?

**Comment by Nekto89** [comment]
It looks like a feature that I need but it doesn't work for some reason (using conan 1.35.0)

conanfile.py based on documentation

```
from conans import ConanFile
from conan.tools.cmake import CMake, CMakeToolchain, CMakeDeps

class App(ConanFile):
    settings = "os", "arch", "compiler", "build_type"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": False, "fPIC": True}
    exports_sources = "CMakeLists.txt"

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()
        deps = CMakeDeps(self)
        deps.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()
```

CMakeLists.txt
```
cmake_minimum_required(VERSION 3.20.1)

project(test_package LANGUAGES CXX)

message(WARNING ${MSVC_VERSION})
```

command
`conan create . test_package/0.1@ --settings arch=x86_64 --settings build_type=RelWithDebInfo --settings compiler=msvc --settings compiler.version=19.26 --settings compiler.runtime=dynamic --settings compiler.runtime_type=Release --settings compiler.cppstd=17`

leads to
```
...
-- The CXX compiler identification is MSVC 19.28.29913.0
...
CMake Warning at CMakeLists.txt:5 (message):
  1928
```

I was expecting to get same result as from
```
cmake -G"Visual Studio 16 2019" -Tversion=14.26 ../conan_test
...
-- The CXX compiler identification is MSVC 19.26.28808.1
...
CMake Warning at CMakeLists.txt:5 (message):
  1926
```

Am I missing something? Do I need to additionally enable some option in my conan configuration?

**Comment by memsharded** [comment]
Hi @Nekto89 

This is good info. The toolchain so far do not implement the specific version selection, only the toolset.
Lets see if we can add this to it. I would really like to add it to the generated ``conan_toolchain.cmake`` file, and not the command line, do you know if this is possible?

For some reason (https://cmake.org/cmake/help/latest/variable/CMAKE_GENERATOR_TOOLSET.html), it seems that the version information can only be passed in command line ``-Tversion``, and the variables described in the help are mostly to read values.

**Comment by Nekto89** [comment]
I've tried adding 4 lines with 1 hardcoded version to end of this method in my local 1.35.0 and it seems to be working so probably all mapping from compiler.version to CMAKE_GENERATOR_TOOLSET should be done there
https://github.com/conan-io/conan/blob/663b24cd7a6105119129bb1684f5cf8418c461b6/conan/tools/cmake/generic.py#L18

```
elif compiler == "msvc":
        compiler_version = settings.get_safe("compiler.version")
        if compiler_version == "19.26":
            return "version=14.26"
```

**Comment by Nekto89** [comment]
> For some reason (https://cmake.org/cmake/help/latest/variable/CMAKE_GENERATOR_TOOLSET.html), it seems that the version information can only be passed in command line `-Tversion`, and the variables described in the help are mostly to read values.

Documentation states "A toolchain file specified by the CMAKE_TOOLCHAIN_FILE variable may initialize CMAKE_GENERATOR_TOOLSET. Once a given build tree has been initialized with a particular value for this variable, changing the value has undefined behavior." so it's ok to pass toolset through toochain file.

**Comment by memsharded** [comment]
Yeah, I was trying to set CMAKE_VS_ vars, I thought that was the way. Didn't know that ``version=xxx`` was allowed syntax for CMAKE_GENERATOR_TOOLSET, this is good info, thanks, I will keep investigating in this line.

**Comment by memsharded** [comment]
I am proposing this PR for next release: https://github.com/conan-io/conan/pull/8800. If you want to have a look or run it from sources, that would be great.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/8800 merged, will be in 1.36
